Add MkDocs format support

giflw opened the pull request:
Add mkdocs single and multipage, plain and with-help, to generator

"MkDocs is a fast, simple and downright gorgeous static site generator that's geared towards building project documentation. Documentation source files are written in Markdown, and configured with a single YAML configuration file. Start by reading the [introductory tutorial](https://www.mkdocs.org/getting-started/), then check the [User Guide](https://www.mkdocs.org/user-guide/) for more information."

MkDocs uses a python parser with support for some PHP Extras features as well lots of features similar to asciidoc, beeing a powerfull markdown documentation builder.

raifdmueller reviewed (approved):
## PR Review: Add MkDocs format support

### Summary
This PR adds excellent support for MkDocs format to the arc42-generator, implementing both single-page and multi-page variants. MkDocs is a valuable addition as it's becoming increasingly popular for technical documentation.

### Strengths
1. **Well-integrated:** The implementation follows existing project patterns and architecture
2. **Clean code:** Well-structured and easy to understand
3. **Proper customization:** Correctly handles MkDocs-specific requirements (docs subdirectory, mkdocs.yml)
4. **Image path handling:** Good adaptation of image paths for MkDocs structure

### Suggestions for enhancement
While the implementation is solid, here are some potential improvements to consider:

1. **MkDocs configuration:** The generated mkdocs.yml is quite minimal. Consider adding:
   - Theme configuration (material theme is popular)
   - Site description and additional metadata
   - More markdown extension options that arc42 could benefit from

2. **Navigation structure:** The multi-page implementation creates a basic index with links, but MkDocs supports hierarchical navigation which could better represent arc42's structure.

3. **Documentation:** It would be helpful to add usage instructions for the new MkDocs formats to the README.

4. **Pandoc parameters:** The current implementation uses 'markdown_phpextra'. It might be worth testing if 'gfm' (GitHub Flavored Markdown) provides better results with MkDocs.

### Conclusion
This is a valuable addition that extends arc42-generator's capabilities to include a popular documentation format. The implementation is clean and follows existing patterns. Approving with suggestions for future enhancements.

raifdmueller commented on subBuild.gradle:
```diff
@@ -285,6 +286,76 @@ task convert2GitHubMarkdownMP (
+            '-t','markdown_phpextra',
+            '-o',localBuildDir.mkdocs+'/docs/index.md',
+        localBuildDir.docbook+'/arc42-template.xml']
+}
```
For the multi-page version, you might want to consider using MkDocs' hierarchical navigation instead of the current approach. Here's how you could structure the `mkdocs.yml` nav section to better represent arc42's structure:

```yaml
nav:
  - Home: index.md
  - Introduction and Goals: 01_introduction_and_goals.md
  - Architecture Constraints: 02_architecture_constraints.md
  - System Scope and Context: 03_system_scope_and_context.md
  - Solution Strategy: 04_solution_strategy.md
  - Building Block View:
      - Level 1: 05_building_block_view/01_level_1.md
      - Level 2: 05_building_block_view/02_level_2.md
      - Level 3: 05_building_block_view/03_level_3.md
  - Runtime View: 06_runtime_view.md
  - Deployment View: 07_deployment_view.md
  - Cross-cutting Concepts: 08_concepts.md
  - Architecture Decisions: 09_architecture_decisions.md
  - Quality Requirements: 10_quality_requirements.md
  - Risks and Technical Debt: 11_technical_risks.md
  - Glossary: 12_glossary.md
```

This would create a more structured navigation experience for users of the generated documentation.

giflw replied:
On single page: hided navigation and kept toc on right side addind automatic navigation

On multi page: kept both navigation and toc automatic as material theme adds proper navigation

raifdmueller commented on subBuild.gradle:
```diff
@@ -285,6 +286,76 @@ task convert2GitHubMarkdownMP (
+                new File(localBuildDir.mkdocsMP + '/mkdocs.yml')
+                    .write("""  |site_name: Arc42 Template
+                                |nav:
+                                |    - index.md
```
Regarding the Pandoc conversion parameters, you might want to test using `gfm` (GitHub Flavored Markdown) instead of `markdown_phpextra`. While `markdown_phpextra` provides good extensions, `gfm` might offer better compatibility with MkDocs as both are designed with modern documentation needs in mind. Here's how you could modify this line:

```groovy
args = ['-r','docbook',
        '-t','gfm',
        '-o',localBuildDir.mkdocs+'/docs/index.md',
    localBuildDir.docbook+'/arc42-template.xml']
```

You could also consider adding some additional Pandoc options to improve the output quality, such as:

```groovy
args = ['-r','docbook',
        '-t','gfm',
        '--toc',                   // Include table of contents
        '--wrap=none',             // No text wrapping
        '--reference-links',       // Use reference-style links
        '-o',localBuildDir.mkdocs+'/docs/index.md',
    localBuildDir.docbook+'/arc42-template.xml']
```

giflw replied:
Added extra options, but kept markdown_phpextra as gfm generates some artifacts in resulting mkdocs html and besides that, mkdocs uses a python parser that uses tables as defined in PHP Markdown Extra

raifdmueller commented on subBuild.gradle:
```diff
@@ -285,6 +286,76 @@ task convert2GitHubMarkdownMP (
+            '-t','markdown_phpextra',
+            '-o',localBuildDir.mkdocs+'/docs/index.md',
+        localBuildDir.docbook+'/arc42-template.xml']
+}
```
The MkDocs configuration is quite minimal currently. I'd recommend enhancing it with a theme (Material is very popular for documentation) and more configuration options. Here's an enhanced example:

```groovy
new File(localBuildDir.mkdocs + '/mkdocs.yml')
    .write("""  |site_name: Arc42 Template
                |site_description: Software architecture documentation with arc42
                |
                |theme:
                |  name: material
                |  features:
                |    - navigation.instant
                |    - navigation.tracking
                |    - navigation.expand
                |    - navigation.indexes
                |    - toc.follow
                |
                |nav:
                |    - index.md
                |
                |markdown_extensions:
                |    - extra
                |    - admonition
                |    - attr_list
                |    - def_list
                |    - footnotes
                |    - meta
                |    - toc:
                |        permalink: true
                |
                |plugins:
                |    - search
                |""".stripMargin())
```

This would provide a much nicer documentation experience with better navigation, search capabilities, and support for additional markdown features useful for architecture documentation.

giflw replied:
Configuration enhanced

rdmueller commented:
Since my knowledge about MKDocs is limited, I asked Claude Sonnet 3.7 to create a review. Do the comments make sense?

giflw commented:
I've updated from master and build is not working on Java 17, so I've used 11.

MP generation is not working for me too, so I've changed to process call directly in task. This should be fixed but I'm not used to gradle.
